llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```diff
@@ -59,32 +67,120 @@ void checkKernel(const Function *F, const CallGraphTy &CG) {
   checkKernelImpl(F, CG, CallStack);
 }
 
+void computeFunctionToKernelsMappingImpl(Function *Kernel, const Function *F,
+                                         const CallGraphTy &CG,
+                                         FuncToFuncMapTy &Mapping) {
+  CallGraphTy::const_iterator It = CG.find(F);
+  // It could be that the function itself is a leaf and doesn't call anything
+  if (It == CG.end())
+    return;
+
+  Mapping[F].insert(Kernel);
+
+  const SmallPtrSet<Value *, 8> &Callees = It->getSecond();
+  for (const Value *V : Callees)
+    if (auto *Callee = dyn_cast<Function>(V))
+      computeFunctionToKernelsMappingImpl(Kernel, Callee, CG, Mapping);
+}
+
+void computeFunctionToKernelsMapping(Function *Kernel, const CallGraphTy &CG,
+                                     FuncToFuncMapTy &Mapping) {
+  // For simplicity we also consider a kernel to be using itself
+  Mapping[Kernel].insert(Kernel);
+
+  CallGraphTy::const_iterator It = CG.find(Kernel);
+  // It could be that the kernel doesn't call anything
+  if (It == CG.end())
+    return;
+
+  const SmallPtrSet<Value *, 8> &Callees = It->getSecond();
+  for (const Value *V : Callees) {
+    auto *Callee = dyn_cast<Function>(V);
+    if (!Callee)
+      continue;
+    Mapping[Callee].insert(Kernel);
+    computeFunctionToKernelsMappingImpl(Kernel, Callee, CG, Mapping);
+  }
+}
+
+void collectVTablesThatUseFunction(
+    const Value *V, SmallVectorImpl<const GlobalVariable *> &VTables) {
+  for (const auto *U : V->users()) {
+    // GlobalVariable is also a constant
+    if (const auto *GV = dyn_cast<GlobalVariable>(U)) {
+      // Core SYCL specification prohibit ODR use of non-const global variables
```

Comment:
```suggestion
      // The Core SYCL specification prohibist ODR use of non-const global variables
```

Reply:
sorry, its `prohibits`, typo'd the typo fix suggestion

Reply:
Applied in 876c8a22f8c12191d4d5b6361ef9c826153c226a